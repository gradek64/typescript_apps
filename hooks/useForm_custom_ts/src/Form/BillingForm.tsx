import React from 'react';

// custom;
import userForm from '../custom/useForm';
import {
  billingFormSchema,
  dropDownPaymentMethod,
} from '../formSchemas/billingFormSchema';
import './styles.scss';

interface Iprops {
  id: string;
}

const AccountForm: React.FC<Iprops> = (props: Iprops) => {
  /* const [name, changeName] = useState<string>('');
  const [surname, changeSurname] = useState<string>(''); */
  const register = () => {
    console.log(':::Account object:::', values);
  };

  const [values, changeValues, validateValues, handleSubmit] = userForm(
    props.id,
    register,
    billingFormSchema
  );

  return (
    <form onSubmit={handleSubmit} id={props.id}>
      <label htmlFor="surname">surname</label>
      <input
        type="text"
        name="surname"
        className={values.surname && values.surname.error && 'error'}
        //value={values.surname ? values.surname.value : ''}
        // no need for above check if values.surname is defined as in useForm hook
        value={values.surname.value}
        onChange={changeValues}
        onBlur={validateValues}
        data-error={values.surname && values.surname.error}
      />
      {values.surname && values.surname.error ? (
        <span className={'errorText'}>{values.surname.error}</span>
      ) : null}
      <label htmlFor="name">name</label>
      <input
        type="text"
        name="name"
        className={values.name && values.name.error && 'error'}
        value={values.name.value}
        onChange={changeValues}
        onBlur={validateValues}
        data-error={values.name && values.name.error}
      />
      {values.name && values.name.error ? (
        <span className={'errorText'}>{values.name.error}</span>
      ) : null}
      <label htmlFor="paymentMethod">payment Method</label>
      <select
        name="paymentMethod"
        className={
          values.paymentMethod && values.paymentMethod.error && 'error'
        }
        defaultValue={values.paymentMethod.value}
        //onBlur only work when you select option so no point
        //to validate on that
        onChange={changeValues}
        data-error={values.paymentMethod && values.paymentMethod.error}
      >
        <option style={{ color: 'gray' }} value="" disabled>
          pick a payment method
        </option>
        {dropDownPaymentMethod.map((option, index) => (
          <option value={option.value} key={option.label + index}>
            {option.label}
          </option>
        ))}
      </select>
      {values.paymentMethod && values.paymentMethod.error ? (
        <span className={'errorText'}>{values.paymentMethod.error}</span>
      ) : null}
      <label htmlFor="accountNumber">accountNumber</label>
      <input
        type="number"
        className={
          values.accountNumber && values.accountNumber.error && 'error'
        }
        name="accountNumber"
        value={values.accountNumber.value}
        onChange={changeValues}
        onBlur={validateValues}
        data-error={values.accountNumber && values.accountNumber.error}
      />
      {values.accountNumber && values.accountNumber.error ? (
        <span className={'errorText'}>{values.accountNumber.error}</span>
      ) : null}
      <input type="submit" value="submit" disabled={false} />
    </form>
  );
};

export default AccountForm;
