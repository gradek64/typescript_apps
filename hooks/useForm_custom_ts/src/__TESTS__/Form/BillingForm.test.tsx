import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import BillingForm from '../../Form/BillingForm';

const initialProps = {
  id: 'BillingForm',
};

describe('should render correclty with initial state', () => {
  afterEach(cleanup);
  it('should render and take a snapshot', () => {
    const { asFragment } = render(<BillingForm {...initialProps} />);
    const component = asFragment();
    expect(component).toMatchSnapshot();
  });
});
