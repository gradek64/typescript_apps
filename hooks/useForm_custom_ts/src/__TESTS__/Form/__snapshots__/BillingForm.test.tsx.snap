// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correclty with initial state should render and take a snapshot 1`] = `
<DocumentFragment>
  <form
    id="BillingForm"
  >
    <label
      for="surname"
    >
      surname
    </label>
    <input
      class=""
      data-error=""
      name="surname"
      type="text"
      value=""
    />
    <label
      for="name"
    >
      name
    </label>
    <input
      class=""
      data-error=""
      name="name"
      type="text"
      value=""
    />
    <label
      for="paymentMethod"
    >
      payment Method
    </label>
    <select
      class=""
      data-error=""
      name="paymentMethod"
    >
      <option
        disabled=""
        selected=""
        style="color: gray;"
        value=""
      >
        pick a payment method
      </option>
      <option
        value="visa card"
      >
        visa card
      </option>
      <option
        value="mastercard"
      >
        mastercard
      </option>
      <option
        value="paypal"
      >
        paypal
      </option>
      <option
        value="klarna"
      >
        klarna
      </option>
    </select>
    <label
      for="accountNumber"
    >
      accountNumber
    </label>
    <input
      class=""
      data-error=""
      name="accountNumber"
      type="number"
      value=""
    />
    <input
      type="submit"
      value="submit"
    />
  </form>
</DocumentFragment>
`;
